lib/posix/pthread.c

```diff
@@ -283,6 +300,50 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
 	CODE_UNREACHABLE;
 }
 
+static void posix_thread_recycle(void)
+{
+	k_spinlock_key_t key;
+	struct posix_thread *t;
+	struct posix_thread *safe_t;
+	sys_dlist_t recyclables = SYS_DLIST_STATIC_INIT(&recyclables);
+
+	key = k_spin_lock(&pthread_pool_lock);
+	SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&done_q, t, safe_t, q_node) {
+		if (t->detachstate == PTHREAD_CREATE_JOINABLE) {
+			/* thread has not been joined yet */
+			continue;
+		}
+
+		sys_dlist_remove(&t->q_node);
+		sys_dlist_append(&recyclables, &t->q_node);
+	}
+	k_spin_unlock(&pthread_pool_lock, key);
+
+	if (sys_dlist_is_empty(&recyclables)) {
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_DYNAMIC_THREAD)) {
+		SYS_DLIST_FOR_EACH_CONTAINER(&recyclables, t, q_node) {
+			if (t->dynamic_stack != NULL) {
+				if (k_is_user_context()) {
+					(void)k_thread_stack_free(t->dynamic_stack);
+				} else {
+					(void)z_impl_k_thread_stack_free(t->dynamic_stack);
+				}
```

Comment:
You don't need to duplicate it. Just call `k_thread_stack_free`, the userspace already checks if is being called from userspace or not.